### ISSUE TITLE
TLS/SSL settings per the README does not work using docker compose

### ISSUE THREAD

**Issue by kwill4026** [bitnami/bitnami-docker-mongodb#173]
**Describe the bug**:

This ticket/bug is separate for the other one I have which involves me running into this issue with the heln chart setup as well as including some other settings. 
My problem here is that I am simply trying to start the cluster with **TLS/SSL** working via the **MONGODB_EXTRA_FLAGS** and **MONGODB_CLIENT_EXTRA_FLAGS** using docker compose this time. 

**Version of Docker **:

```
Client: Docker Engine - Community
 Version:           19.03.1
Server: Docker Engine - Community
  Version:          19.03.1
```

Docker Compose Version :
```
docker-compose version 1.18.0, build 8dd22a9, docker-py version: 2.6.1
```

**Which chart**: 
```
bitnami/bitnami-docker-mongodb
```

**What happened**:

Using docker-compose to start the **bitnami/bitnami-docker-mongodb** works fine as the default with no TLS/SSL. However attempting to enable TLS/SSL between the nodes in the cluster, as well as between mongo clients and nodes, by setting the **MONGODB_EXTRA_FLAGS** and **MONGODB_CLIENT_EXTRA_FLAGS** environment variables, together with the correct **MONGODB_ADVERTISED_HOSTNAME** does not work. I also generated the CA cert and pem files ahead of time. When looking at the logs of the pod I get:
```
INFO  ==> ** Starting MongoDB **
2019-09-13T17:49:10.705+0000 I STORAGE  [main] Max cache overflow file size custom option: 0
BadValue: unsupported value for sslMode requireSSL --sslPEMKeyFile=/certs/mongodb-primary.pem --sslClusterFile=/certs/mongodb-primary.pem --sslCAFile=/certs/mongoCA.crt
try '/opt/bitnami/mongodb/bin/mongod --help' for more information
```

**What you expected to happen**:

I expect the `mongod` to be enabled to run with TLS/SSL because of the settings I enabled as well as the certs I generated. 

**How to reproduce it** (as minimally and precisely as possible):** 

Generate PEM certificates as required by Mongo by using openSSL. Then mount the directory you placed those certs in to a directory in the container. Then specify in the docker-compose file the **MONGODB_EXTRA_FLAGS** AND **MONGODB_CLIENT_EXTRA_FLAGS** settings as so:

```yaml
version: '2'

services:
  mongodb-primary:
    image: 'bitnami/mongodb:latest'
    environment:
      - MONGODB_ADVERTISED_HOSTNAME=mongodb-primary
      - MONGODB_REPLICA_SET_MODE=primary
      - MONGODB_ROOT_PASSWORD=password123
      - MONGODB_REPLICA_SET_KEY=key12345
      - MONGODB_EXTRA_FLAGS=--sslMode=requireSSL --sslPEMKeyFile=/certs/mongodb-primary.pem --sslClusterFile=/certs/mongodb-primary.pem --sslCAFile=/certs/mongoCA.crt
      - MONGODB_CLIENT_EXTRA_FLAGS=--ssl --sslPEMKeyFile=/certs/mongodb-primary.pem --sslCAFile=/certs/mongoCA.crt

    ports:
      - "27017:27017"
    volumes:
      - 'mongodb_master_data:/bitnami'
      - './certificates:/certs'
      - '/home/kwill4026/compose_files/bitnami-docker-mongodb/mongodb.conf:/opt/bitnami/mongodb/conf/mongodb.conf'
  mongodb-secondary:
    image: 'bitnami/mongodb:latest'
    depends_on:
      - mongodb-primary
    environment:
      - MONGODB_ADVERTISED_HOSTNAME=mongodb-secondary
      - MONGODB_REPLICA_SET_MODE=secondary
      - MONGODB_PRIMARY_HOST=mongodb-primary
      - MONGODB_PRIMARY_PORT_NUMBER=27017
      - MONGODB_PRIMARY_ROOT_PASSWORD=password123
      - MONGODB_REPLICA_SET_KEY=key12345
      - MONGODB_EXTRA_FLAGS=--sslMode=requireSSL --sslPEMKeyFile=/certs/mongodb-secondary.pem --sslClusterFile=/certs/mongodb-secondary.pem --sslCAFile=/certs/mongoCA.crt
      - MONGODB_CLIENT_EXTRA_FLAGS=--ssl --sslPEMKeyFile=/certs/mongodb-secondary.pem --sslCAFile=/certs/mongoCA.crt

  mongodb-arbiter:
    image: 'bitnami/mongodb:latest'
    depends_on:
      - mongodb-primary
    environment:
      - MONGODB_ADVERTISED_HOSTNAME=mongodb-arbiter
      - MONGODB_REPLICA_SET_MODE=arbiter
      - MONGODB_PRIMARY_HOST=mongodb-primary
      - MONGODB_PRIMARY_PORT_NUMBER=27017
      - MONGODB_PRIMARY_ROOT_PASSWORD=password123
      - MONGODB_REPLICA_SET_KEY=key12345

volumes:
  mongodb_master_data:
  certificates:
    driver: local
```

**Anything else we need to know**:

According to the `bitnami/bitnami-docker-mongodb` located [here](https://github.com/bitnami/bitnami-docker-mongodb):

> This container supports enabling SSL/TLS between nodes in the cluster, as well as between mongo clients and nodes, by setting the MONGODB_EXTRA_FLAGS and MONGODB_CLIENT_EXTRA_FLAGS environment variables or by specifying it in the mongo configmap settings. 

I have done this but get dinged on the --sslMode in which I have tried all the values **disabled,allowSSL,requireSSL, and preferSSL**

**Comment by juan131** [comment]
Hi @kwill4026 

Did you recently downloaded the latest version of `bitnami/mongodb`? I think this issue is already solved in the latest image but you're probably using some image you already had in your cache. You can try it by running the commands:

```console
$ docker-compose down -v --rmi all
$ docker-compose up
```

Please do not hesitate to let us know if the error persists in the latest image.

**Comment by kwill4026** [comment]
Hello @juan131 , 
Which is the latest version of bitnami/mongodb? Is it simple just the bitnami/mongodb:latest ? In which from reading the digest of this from docker hub seems to be version 4.0.12-debian-9-r38 . 
Or is the latest version some form of 4.2* variant?

**Comment by juan131** [comment]
You can find the information about the **latest** tag and the branch which it points to in the link below:

https://github.com/bitnami/bitnami-docker-mongodb#supported-tags-and-respective-dockerfile-links

**Comment by kwill4026** [comment]
This works fine for the docker-compose file setup but not for helm charts. For some reason once I specify parameters such as 
```
 -  "--wiredTigerCacheSizeGB=5"
 -  "--port=27017"
 -  "--bind_ip=0.0.0.0" 
```
it works fine with no problem. But specifying ssl parameters such as
 ```
 -  "--wiredTigerCacheSizeGB=5"
 -  "--port=27017"
 -  "--sslMode=preferSSL"
```
the pods fail to run. 
or just even a single parameter
```
-  "--sslAllowConnectionsWithoutCertificates"
```
 causes the pod to fail. I will have to open another ticket with the helm chart people about this.

**Comment by juan131** [comment]
Hi @kwill4026 

I guess you're talking about this ticket: https://github.com/helm/charts/issues/17472

>  I will have to open another ticket with the helm chart people about this.

It turns out that we (Bitnami) are also the maintainers of the chart.

**Comment by kwill4026** [comment]
Yes @juan131. Ok good to know

**Comment by ThomasLachaux** [comment]
I have exactly the same problem but I am not using helm. 
I have the latest docker version, docker-compose version and bitnami image

Does others still encounter this issue ?

**Comment by carrodher** [comment]
In order to try to reproduce the issue, can you provide us the docker-compose you're using and any extra configuration you're using?

It shouldn't be happening with the recent versions of the image

**Comment by ThomasLachaux** [comment]
Hello !

It is strange because I have followed exactly the README.

**You can find the project here if you want to test it
[Bitnami docker.zip](https://github.com/bitnami/bitnami-docker-mongodb/files/5067861/Bitnami.docker.zip)**



Docker version 19.03.12, build 48a66213fe
docker-compose version 1.26.2, build eefe0d31
Working on MacOS 10.15.6
Working with image bitnami/mongodb:latest (848f7fa90d47)

`docker-compose.yml`
```
version: '2.4'

services:
  mongo-1:
    container_name: mongo-1
    image: 'bitnami/mongodb'
    environment:
      MONGODB_ADVERTISED_HOSTNAME: mongo-1
      MONGODB_REPLICA_SET_MODE: primary
      MONGODB_ROOT_PASSWORD: root
      MONGODB_REPLICA_SET_KEY: JvUfTK5GZhHdX7K3
      MONGODB_EXTRA_FLAGS: "--sslMode=requireSSL --sslPEMKeyFile=/certificates/mongodb.pem --sslClusterFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt"
      MONGODB_CLIENT_EXTRA_FLAGS: "--ssl --sslPEMKeyFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt"
      MONGODB_SYSTEM_LOG_VERBOSITY: 3
      BITNAMI_DEBUG: "true"
    volumes:
      - './certs/mongo-1.pem:/certificates/mongodb.pem'
      - './certs/ca.crt:/certificates/mongoCA.crt'

  mongo-2:
    container_name: mongo-2
    image: 'bitnami/mongodb'
    depends_on:
      - mongo-1
    environment:
      - MONGODB_ADVERTISED_HOSTNAME=mongo-2
      - MONGODB_REPLICA_SET_MODE=secondary
      - MONGODB_INITIAL_PRIMARY_HOST=mongo-1
      - MONGODB_INITIAL_PRIMARY_ROOT_PASSWORD=root
      - MONGODB_REPLICA_SET_KEY=JvUfTK5GZhHdX7K3
      - MONGODB_EXTRA_FLAGS=--sslMode=requireSSL --sslPEMKeyFile=/certificates/mongodb.pem --sslClusterFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt
      - MONGODB_CLIENT_EXTRA_FLAGS=--ssl --sslPEMKeyFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt
    volumes:
     - './certs/mongo-2.pem:/certificates/mongodb.pem'
     - './certs/ca.crt:/certificates/mongoCA.crt'

  mongo-3:
    container_name: mongo-3
    image: 'bitnami/mongodb'
    depends_on:
      - mongo-1
    environment:
      - MONGODB_ADVERTISED_HOSTNAME=mongo-3
      - MONGODB_REPLICA_SET_MODE=secondary
      - MONGODB_INITIAL_PRIMARY_HOST=mongo-1
      - MONGODB_INITIAL_PRIMARY_ROOT_PASSWORD=root
      - MONGODB_REPLICA_SET_KEY=JvUfTK5GZhHdX7K3
      - MONGODB_EXTRA_FLAGS=--sslMode=requireSSL --sslPEMKeyFile=/certificates/mongodb.pem --sslClusterFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt
      - MONGODB_CLIENT_EXTRA_FLAGS=--ssl --sslPEMKeyFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt
    volumes:
      - './certs/mongo-3.pem:/certificates/mongodb.pem'
      - './certs/ca.crt:/certificates/mongoCA.crt'
```


*Logs on the primary node:*
```
mongodb 08:03:23.33
mongodb 08:03:23.34 Welcome to the Bitnami mongodb container
mongodb 08:03:23.34 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-mongodb
mongodb 08:03:23.35 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-mongodb/issues
mongodb 08:03:23.35
mongodb 08:03:23.37 INFO  ==> ** Starting MongoDB setup **
mongodb 08:03:23.41 INFO  ==> Validating settings in MONGODB_* env vars...
mongodb 08:03:23.46 INFO  ==> Initializing MongoDB...
mongodb 08:03:23.52 INFO  ==> Deploying MongoDB from scratch...
mongodb 08:03:23.54 DEBUG ==> Starting MongoDB in background...
2020-08-13T08:03:23.602+0000 W  CONTROL  [main] Option: sslMode is deprecated. Please use tlsMode instead.
2020-08-13T08:03:23.602+0000 W  CONTROL  [main] Option: sslPEMKeyFile is deprecated. Please use tlsCertificateKeyFile instead.
2020-08-13T08:03:23.602+0000 W  CONTROL  [main] Option: sslClusterFile is deprecated. Please use tlsClusterFile instead.
2020-08-13T08:03:23.602+0000 W  CONTROL  [main] Option: sslCAFile is deprecated. Please use tlsCAFile instead.
about to fork child process, waiting until server is ready for connections.
forked process: 39
child process started successfully, parent exiting
mongodb 08:03:24.99 INFO  ==> Creating users...
mongodb 08:03:25.00 INFO  ==> Creating root user...
mongodb 08:03:25.17 INFO  ==> Stopping MongoDB...
```
and then exits with code 1

*Logs on the secondary node*
```
mongodb 08:03:24.13
mongodb 08:03:24.13 Welcome to the Bitnami mongodb container
mongodb 08:03:24.13 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-mongodb
mongodb 08:03:24.13 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-mongodb/issues
mongodb 08:03:24.14
mongodb 08:03:24.14 INFO  ==> ** Starting MongoDB setup **
mongodb 08:03:24.26 INFO  ==> Validating settings in MONGODB_* env vars...
mongodb 08:03:24.36 INFO  ==> Initializing MongoDB...
mongodb 08:03:24.38 INFO  ==> Deploying MongoDB from scratch...
mongodb 08:03:25.43 INFO  ==> Creating users...
mongodb 08:03:25.43 INFO  ==> Users created
mongodb 08:03:25.44 INFO  ==> Writing keyfile for replica set authentication...
mongodb 08:03:25.47 INFO  ==> Configuring MongoDB replica set...
mongodb 08:03:25.47 INFO  ==> Stopping MongoDB...
mongodb 08:03:27.88 INFO  ==> Trying to connect to MongoDB server mongo-1...
cannot resolve host "mongo-1": lookup mongo-1 on 127.0.0.11:53: no such host
cannot resolve host "mongo-1": lookup mongo-1 on 127.0.0.11:53: no such host
cannot resolve host "mongo-1": lookup mongo-1 on 127.0.0.11:53: no such host
*More of these messages...*
mongodb 08:06:23.29 ERROR ==> Unable to connect to host mongo-1
mongodb 08:06:23.29 INFO  ==> Stopping MongoDB...
```
and then exits with code 1

Have a nice day !

**Comment by carrodher** [comment]
Hi,

I did some changes to your docker-compose.yaml and it seems everything is working.

- First, it seems the env. variables in the primary node are not property set:
```diff
<       MONGODB_ADVERTISED_HOSTNAME: mongo-1
<       MONGODB_REPLICA_SET_MODE: primary
<       MONGODB_ROOT_PASSWORD: root
<       MONGODB_REPLICA_SET_KEY: JvUfTK5GZhHdX7K3
<       MONGODB_EXTRA_FLAGS: "--sslMode=requireSSL --sslPEMKeyFile=/certificates/mongodb.pem --sslClusterFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt"
<       MONGODB_CLIENT_EXTRA_FLAGS: "--ssl --sslPEMKeyFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt"
<       MONGODB_SYSTEM_LOG_VERBOSITY: 3
<       BITNAMI_DEBUG: "true"
---
>       - MONGODB_ADVERTISED_HOSTNAME=mongo-1
>       - MONGODB_REPLICA_SET_MODE=primary
>       - MONGODB_ROOT_PASSWORD=root
>       - MONGODB_REPLICA_SET_KEY=JvUfTK5GZhHdX7K3
>       - MONGODB_EXTRA_FLAGS=--sslMode=requireSSL --sslPEMKeyFile=/certificates/mongodb.pem --sslClusterFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt
>       - MONGODB_CLIENT_EXTRA_FLAGS=--ssl --sslPEMKeyFile=/certificates/mongodb.pem --sslCAFile=/certificates/mongoCA.crt
>       - MONGODB_SYSTEM_LOG_VERBOSITY=3
>       - BITNAMI_DEBUG=true
```
please note the `KEY: value` VS `- KEY=value` and also the removal of some backquotes.

- Then, when this section in the README was created, the latest MongoDB version was 4.1. Now, when using `image: 'bitnami/mongodb'` you're pulling the latest one (MongoDB 4.2) and some parameters were renamed, so I also adapted the docker-compose.yaml in order to use the previous image and the env. variables used with that image:
```diff
6c6
<     image: 'bitnami/mongodb'
---
>     image: 'bitnami/mongodb:4.1'

22c22
<     image: 'bitnami/mongodb'
---
>     image: 'bitnami/mongodb:4.1'
28,29c28,29
<       - MONGODB_INITIAL_PRIMARY_HOST=mongo-1
<       - MONGODB_INITIAL_PRIMARY_ROOT_PASSWORD=root
---
>       - MONGODB_PRIMARY_HOST=mongo-1
>       - MONGODB_PRIMARY_ROOT_PASSWORD=root

39c39
<     image: 'bitnami/mongodb'
---
>     image: 'bitnami/mongodb:4.1'
45,46c45,46
<       - MONGODB_INITIAL_PRIMARY_HOST=mongo-1
<       - MONGODB_INITIAL_PRIMARY_ROOT_PASSWORD=root
---
>       - MONGODB_PRIMARY_HOST=mongo-1
>       - MONGODB_PRIMARY_ROOT_PASSWORD=root
```
basically set a 4.1 version and replace `MONGODB_INITIAL_PRIMARY_*` by `MONGODB_PRIMARY_*`.

Maybe it is needed to adapt some env. variable of parameters if you want to use latest MongoDB (4.2)

**Comment by ThomasLachaux** [comment]
Oh my god, you are so awesome ! 

Thank you so much for your help ♥️

**Comment by kwill4026** [comment]
So yes its definitely something wrong with the 4.2 tag or like @carrodher mentioned its missing some env variable or parameter in our docker-compose.yaml file for the TLS settings when the 4.2 tag is used. I say this because it works fine when using the 4.1 tag or even the 4.2.0 tag. However when I used 4.2 or even the latest tag (which is really 4.2.8-debian-10-r46) it does not work.

**Comment by carrodher** [comment]
Hi, I just changed the flags in the README, since it seems they are deprecated and the following warning was appearing:
```
mongo-1    | 2020-08-14T09:07:52.704+0000 W  CONTROL  [main] Option: sslMode is deprecated. Please use tlsMode instead.
mongo-1    | 2020-08-14T09:07:52.704+0000 W  CONTROL  [main] Option: sslPEMKeyFile is deprecated. Please use tlsCertificateKeyFile instead.
mongo-1    | 2020-08-14T09:07:52.704+0000 W  CONTROL  [main] Option: sslClusterFile is deprecated. Please use tlsClusterFile instead.
mongo-1    | 2020-08-14T09:07:52.704+0000 W  CONTROL  [main] Option: sslCAFile is deprecated. Please use tlsCAFile instead.
```
see https://github.com/bitnami/bitnami-docker-mongodb/commit/93448f8acf9361483a81153eff8761b5622e570b

About the issue with the recent versions we are going to investigate it, it's true it is working fine with `4.2.0` but not with `>4.2.1`

**Comment by kwill4026** [comment]
You are correct about the SSL deprecation even though on a 4.2+ deployment it will still take SSL settings but will just display a warning to use TLS settings instead. 
 Let me know what you get on investigating the 4.2.1+ tags. I am having some issues with the TLS settings using the helm deployment and suspect the tag 4.2 and up is causing the errors I am seeing. @carrodher

**Comment by ThomasLachaux** [comment]
Hi !
Any updates ?
Maybe we should reopen the issue/create a new one ^^

**Comment by javsalgar** [comment]
Hi,

Yes, I agree that it we should create a new one just focused on this exact version of MongoDB
